src/ParseFasm.cpp

```diff
@@ -0,0 +1,577 @@
+        /// Tags: annotation (a), annotation name (.), annotation value (=)
+        virtual Any visitAnnotation(
+            FasmParser::AnnotationContext* context) override {
+                std::ostringstream data;
+                data << Str(TAG('.', annotation_name),
+                            context->ANNOTATION_NAME()->getText());
+                if (context->ANNOTATION_VALUE()) {
+                        std::string value =
+                            context->ANNOTATION_VALUE()->getText();
+                        value.erase(0, 1);  /// Convert "value" -> value
+                        value.pop_back();
+                        data << Str(TAG('=', annotation_value), value);
+                }
+                return withHeader(TAG('a', annotation), data.str());
+        }
+
+       private:
+        std::ostream& out;
+};
+
+class FasmErrorListener : public BaseErrorListener {
+       public:
+        virtual void syntaxError(Recognizer* recognizer,
+                                 Token* token,
+                                 size_t line,
+                                 size_t position,
+                                 const std::string& msg,
+                                 std::exception_ptr e) override {
+                throw ParseException{.line = line,
+                                     .position = position,
+                                     .message = std::string(msg)};
+        }
+};
+
+/// Common portion of 'from_string' and 'from_file'.
+/// Consumes an input stream and produces an output stream.
+static void parse_fasm(std::istream& in, std::ostream& out) {
+        ANTLRInputStream stream(in);
+        FasmLexer lexer(&stream);
+        FasmErrorListener errorListener;
+        lexer.removeErrorListeners();
+        lexer.addErrorListener(&errorListener);
+        CommonTokenStream tokens(&lexer);
+        FasmParser parser(&tokens);
+        parser.removeErrorListeners();
+        parser.addErrorListener(&errorListener);
+        auto* tree = parser.fasmFile();
+        FasmParserBaseVisitor(out).visit(tree);
+}
+
+/// Parse the given input string, returning output.
+/// Use hex mode (see above) if hex is true.
+/// Use a callback to avoid copying the result.
+void from_string(const char* in,
+                 bool hex,
+                 void (*ret)(const char* str, size_t)) {
+        hex_mode = hex;
+        std::istringstream input(in);
+        std::ostringstream output;
+
+        try {
+                parse_fasm(input, output);
+                output.put(0);
+                std::string result = output.str();
+                ret(result.c_str(), result.size());
+        } catch (ParseException e) {
+                // Parse failure will throw this exception.
+                // ret() will not be called on parse failure, so
+                // external code can use this to detect the exception.
+                std::cerr << "Parse error at " << e.line << ":" << e.position
```

Comment:
This should be passed to a callback?

Reply:
Same issue with `couldn't open file` in `from_file`?

Reply:
Done.